<?xml version="1.0"?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /opt/ros/kinetic/share/moveit_resources/manipulator_description/urdf/manipulator.urdf | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->

<robot name="manipulator" xmlns:xacro="http://www.ros.org/wiki/xacro">
    <link name="manipulator_link0">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link0.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link0.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="-0.041018 -0.00014 0.049974"/>
            <mass value="0.629769"/>
            <inertia ixx="0.00315" ixy="8.2904e-07" ixz="0.00015" iyy="0.00388" iyz="8.2299e-06" izz="0.004285"/>
        </inertial>
    </link>

    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link0_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link0_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link0"/>
        <child link="manipulator_link0_sc"/>
    </joint>

    <link name="manipulator_link1">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link1.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link1.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="0.003875 0.002081 -0.04762"/>
            <mass value="4.970684"/>
            <inertia ixx="0.70337" ixy="-0.000139" ixz="0.006772" iyy="0.70661" iyz="0.019169" izz="0.009117"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->            
    <link name="manipulator_link1_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link1_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link1"/>
        <child link="manipulator_link1_sc"/>
    </joint>
    <joint name="manipulator_joint1" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
        <origin rpy="0 0 0" xyz="0 0 0.333" />
        <parent link="manipulator_link0" />
        <child link="manipulator_link1" />
        <axis xyz="0 0 1" />
        <limit effort="87" lower="-2.9671" upper="2.9671" velocity="2.3925" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link2">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link2.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link2.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="-0.003141 -0.02872  0.003495"/>
            <mass value="0.646926"/>
            <inertia ixx="0.007962" ixy="-0.003925" ixz="0.010254" iyy="0.02811" iyz="0.000704" izz="0.025995"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link2_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link2_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link2"/>
        <child link="manipulator_link2_sc"/>
    </joint>
    <joint name="manipulator_joint2" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628" />
        <origin rpy="-1.57079632679 0 0" xyz="0 0 0" />
        <parent link="manipulator_link1" />
        <child link="manipulator_link2" />
        <axis xyz="0 0 1" />
        <limit effort="87" lower="-1.8326" upper="1.8326" velocity="2.3925" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link3">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link3.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link3.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="2.7518e-02 3.9252e-02 -6.6502e-02"/>
            <mass value="3.228604"/>
            <inertia ixx="0.037242" ixy="-0.004761" ixz="-0.011396" iyy="0.036155" iyz="-0.012805" izz="0.01083"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link3_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link3_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link3"/>
        <child link="manipulator_link3_sc"/>
    </joint>
    <joint name="manipulator_joint3" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
        <origin rpy="1.57079632679 0 0" xyz="0 -0.316 0" />
        <parent link="manipulator_link2" />
        <child link="manipulator_link3" />
        <axis xyz="0 0 1" />
        <limit effort="87" lower="-2.9671" upper="2.9671" velocity="2.3925" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link4">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link4.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link4.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="-5.317e-02 1.04419e-01 2.7454e-02"/>
            <mass value="3.587895"/>
            <inertia ixx="0.025853" ixy="0.007796" ixz="-0.001332" iyy="0.019552" iyz="0.008641" izz="0.028323"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link4_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link4_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link4"/>
        <child link="manipulator_link4_sc"/>
    </joint>
    <joint name="manipulator_joint4" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="0.0175" />
        <origin rpy="1.57079632679 0 0" xyz="0.0825 0 0" />
        <parent link="manipulator_link3" />
        <child link="manipulator_link4" />
        <axis xyz="0 0 1" />
        <limit effort="87" lower="-3.1416" upper="0.0873" velocity="2.3925" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link5">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link5.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link5.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="-1.1953e-02 4.1065e-02 -3.8437e-02"/>
            <mass value="1.225946"/>
            <inertia ixx="0.035549" ixy="-0.002117" ixz="-0.004037" iyy="0.029474" iyz="0.000229" izz="0.008627"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link5_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link5_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link5"/>
        <child link="manipulator_link5_sc"/>
    </joint>
    <joint name="manipulator_joint5" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
        <origin rpy="-1.57079632679 0 0" xyz="-0.0825 0.384 0" />
        <parent link="manipulator_link4" />
        <child link="manipulator_link5" />
        <axis xyz="0 0 1" />
        <limit effort="12" lower="-2.9671" upper="2.9671" velocity="2.8710" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link6">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link6.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link6.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
            <mass value="1.47"/>
            <inertia ixx="0.0049" ixy="0.0" ixz="0.0" iyy="0.0047" iyz="0.0" izz="0.0036"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link6_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link6_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link6"/>
        <child link="manipulator_link6_sc"/>
    </joint>
    <joint name="manipulator_joint6" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525" />
        <origin rpy="1.57079632679 0 0" xyz="0 0 0" />
        <parent link="manipulator_link5" />
        <child link="manipulator_link6" />
        <axis xyz="0 0 1" />
        <limit effort="12" lower="-0.0873" upper="3.8223" velocity="2.8710" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link7">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/link7.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/link7.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
            <mass value="0.45"/>
            <inertia ixx="0.0035" ixy="0.0" ixz="0.0" iyy="0.0033" iyz="0.0" izz="0.0029"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_link7_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_link7_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_link7"/>
        <child link="manipulator_link7_sc"/>
    </joint>
    <joint name="manipulator_joint7" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973" />
        <origin rpy="1.57079632679 0 0" xyz="0.088 0 0" />
        <parent link="manipulator_link6" />
        <child link="manipulator_link7" />
        <axis xyz="0 0 1" />
        <limit effort="12" lower="-2.9671" upper="2.9671" velocity="2.8710" />
        <dynamics D="1.0" K="7000.0" damping="0.003" friction="0.0" mu_coulomb="0.0" mu_viscous="16.0"/>
    </joint>
    <link name="manipulator_link8" />
    <joint name="manipulator_joint8" type="fixed">
        <origin rpy="0 0 0" xyz="0 0 0.107" />
        <parent link="manipulator_link7" />
        <child link="manipulator_link8" />
        <!-- <axis xyz="0 0 0" /> -->
    </joint>
    <joint name="manipulator_hand_joint" type="fixed">
        <parent link="manipulator_link8" />
        <child link="manipulator_hand" />
        <origin rpy="0 0 -0.785398163397" xyz="0 0 0" />
    </joint>
    <link name="manipulator_hand">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/hand.dae" />
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/collision/hand.stl" />
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0.0 0.0 0.0" xyz="-0.01 0.0 0.03"/>
            <mass value="0.73"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.0025" iyz="0.0" izz="0.0017"/>
        </inertial>
    </link>
    <!-- sub-link defining coarse geometries of real robot's internal self-collision -->
    <link name="manipulator_hand_sc">
    </link>
    <!-- fixed joint between both sub-links -->
    <joint name="manipulator_hand_sc_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0"/>
        <parent link="manipulator_hand"/>
        <child link="manipulator_hand_sc"/>
    </joint>
    <!-- Define the hand_tcp frame -->
    <link name="manipulator_hand_tcp"/>
    <joint name="manipulator_hand_tcp_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.1034"/>
        <parent link="manipulator_hand"/>
        <child link="manipulator_hand_tcp"/>
    </joint>
    <link name="manipulator_leftfinger">
        <visual>
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/finger.dae" />
            </geometry>
        </visual>
        <!-- screw mount -->
        <collision>
            <origin rpy="0.0 0.0 0.0" xyz="0.0 18.5e-3 11.0e-3"/>
            <geometry>
                <box size="22e-3 15e-3 20e-3"/>
            </geometry>
        </collision>
        <!-- cartriage sledge -->
        <collision>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 6.8e-3 2.2e-3"/>
        <geometry>
            <box size="22e-3 8.8e-3 3.8e-3"/>
        </geometry>
        </collision>
        <!-- diagonal finger -->
        <collision>
        <origin rpy="0.5235987755982988 0.0 0.0" xyz="0.0 15.9e-3 28.35e-3"/>
        <geometry>
            <box size="17.5e-3 7e-3 23.5e-3"/>
        </geometry>
        </collision>
        <!-- rubber tip with which to grasp -->
        <collision>
        <origin rpy="0.0 0.0 0.0" xyz="0 7.58e-3 45.25e-3"/>
        <geometry>
            <box size="17.5e-3 15.2e-3 18.5e-3"/>
        </geometry>
        </collision>
        <inertial>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
        <mass value="0.015"/>
        <inertia ixx="2.3749999999999997e-06" ixy="0.0" ixz="0.0" iyy="2.3749999999999997e-06" iyz="0.0" izz="7.5e-07"/>
        </inertial>
    </link>
    <link name="manipulator_rightfinger">
        <visual>
            <origin rpy="0 0 3.14159265359" xyz="0 0 0" />
            <geometry>
                <mesh filename="file://$(find manipulator)/meshes/visual/finger.dae" />
            </geometry>
        </visual>
         <!-- screw mount -->
        <collision>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 -18.5e-3 11e-3"/>
        <geometry>
            <box size="22e-3 15e-3 20e-3"/>
        </geometry>
        </collision>
        <!-- cartriage sledge -->
        <collision>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 -6.8e-3 2.2e-3"/>
        <geometry>
            <box size="22e-3 8.8e-3 3.8e-3"/>
        </geometry>
        </collision>
        <!-- diagonal finger -->
        <collision>
        <origin rpy="-0.5235987755982988 0.0 0.0" xyz="0 -15.9e-3 28.35e-3"/>
        <geometry>
            <box size="17.5e-3 7e-3 23.5e-3"/>
        </geometry>
        </collision>
        <!-- rubber tip with which to grasp -->
        <collision>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 -7.58e-3 45.25e-3"/>
        <geometry>
            <box size="17.5e-3 15.2e-3 18.5e-3"/>
        </geometry>
        </collision>
        <inertial>
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
        <mass value="0.015"/>
        <inertia ixx="2.3749999999999997e-06" ixy="0.0" ixz="0.0" iyy="2.3749999999999997e-06" iyz="0.0" izz="7.5e-07"/>
        </inertial>
    </link>
    <joint name="manipulator_finger_joint1" type="prismatic">
        <parent link="manipulator_hand" />
        <child link="manipulator_leftfinger" />
        <origin rpy="0 0 0" xyz="0 0 0.0584" />
        <axis xyz="0 1 0" />
        <limit effort="100.0" lower="0.0" upper="0.04" velocity="0.2" />
        <dynamics damping="0.3"/>
    </joint>
    <joint name="manipulator_finger_joint2" type="prismatic">
        <parent link="manipulator_hand" />
        <child link="manipulator_rightfinger" />
        <origin rpy="0 0 0" xyz="0 0 0.0584" />
        <axis xyz="0 -1 0" />
        <limit effort="100.0" lower="0.0" upper="0.04" velocity="0.2" />
        <mimic joint="manipulator_finger_joint1" />
        <dynamics damping="0.3"/>
    </joint>
    <link name="world">
    </link>
    <joint name="world_joint" type="fixed">
        <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
        <parent link="world"/>
        <child link="manipulator_link0"/>
    </joint>

    <xacro:macro name="box_inertia" params="m l w h xyz rpy">
        <inertial>
            <origin xyz="${xyz}" rpy="${rpy}" />
            <mass value="${m}" />
            <inertia ixx="${(m/12) * (h*h + l*l)}" ixy="0" ixz="0"
                     iyy="${(m/12) * (w*w + l*l)}" iyz="0"
                     izz="${(m/12) * (w*w + h*h)}" />
        </inertial>
    </xacro:macro>

    <!-- RealSense Camera -->
    <link name="camera_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02505 0.09 0.025"/>
            </geometry>
            <material name="Gazebo/Blue">
                <color rgba="0 0 1 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02505 0.09 0.025"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0001" iyz="0" izz="0.0001"/>
        </inertial>
    </link>

    <joint name="base_camera_joint" type="fixed">
        <parent link="manipulator_hand"/>
        <child link="camera_link"/>
        <origin xyz="0.02 0 0.06" rpy="0 ${-pi/2} 0"/>
    </joint>

    <link name="camera_link_optical"/>

    <joint name="camera_optical_joint" type="fixed">
        <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
        <parent link="camera_link"/>
        <child link="camera_link_optical"/>
    </joint>

    <gazebo reference="camera_link">
        <material>Gazebo/Blue</material>
        <sensor name="camera_sensor" type="depth">
            <update_rate>30.0</update_rate>
            <camera name="camera">
                <horizontal_fov>1.0472</horizontal_fov>
                <image>
                    <width>640</width>
                    <height>480</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>100</far>
                </clip>
                <noise>
                    <type>gaussian</type>
                    <mean>0.0</mean>
                    <stddev>0.007</stddev>
                </noise>
            </camera>
            <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
                <frame_name>camera_link_optical</frame_name>
                <min_depth>0.1</min_depth>
                <max_depth>100.0</max_depth>
            </plugin>
        </sensor>
    </gazebo>
    
    <ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>

        <!-- Arm joints -->
        <joint name="manipulator_joint1">
            <param name="initial_position">0.0</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint2">
            <param name="initial_position">-0.785</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint3">
            <param name="initial_position">0.0</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint4">
            <param name="initial_position">-2.356</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint5">
            <param name="initial_position">0.0</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint6">
            <param name="initial_position">1.571</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
        <joint name="manipulator_joint7">
            <param name="initial_position">0.785</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

        <!-- Gripper joint -->
        <joint name="manipulator_finger_joint1">
            <param name="initial_position">0.0</param>
            <command_interface name="effort"/>
            <command_interface name="velocity"/>
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
    </ros2_control>
    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
            <robot_param>robot_description</robot_param>
            <robot_param_node>robot_state_publisher</robot_param_node>
            <parameters>$(find manipulator)/config/controllers.yaml</parameters>
        </plugin>
    </gazebo>

    <!-- Add Gazebo materials for each link -->
    <gazebo reference="manipulator_link0">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link1">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link2">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link3">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link4">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link5">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link6">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_link7">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_hand">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_leftfinger">
        <material>Gazebo/Grey</material>
    </gazebo>
    <gazebo reference="manipulator_rightfinger">
        <material>Gazebo/Grey</material>
    </gazebo>
</robot>
